src/main/java/frc/robot/poseestimator/GBPoseEstimator.java

```diff
@@ -0,0 +1,228 @@
+		return estimatedPoseAtTimestamp.orElseGet(() -> estimatedPose);
+	}
+
+	@Override
+	public void updateVision(List<VisionObservation> visionObservations) {
+		for (VisionObservation visionObservation : visionObservations) {
+			if (!isObservationTooOld(visionObservation)) {
+				addVisionObservation(visionObservation);
+			}
+		}
+	}
+
+	@Override
+	public void updateOdometry(List<OdometryObservation> odometryObservations) {
+		for (OdometryObservation observation : odometryObservations) {
+			addOdometryObservation(observation);
+		}
+		logEstimatedPose();
+	}
+
+	@Override
+	public void updatePoseEstimator(List<OdometryObservation> odometryObservation, List<VisionObservation> visionObservations) {
+		updateOdometry(odometryObservation);
+		updateVision(visionObservations);
+	}
+
+	private boolean isObservationTooOld(VisionObservation visionObservation) {
+		try {
+			return odometryPoseInterpolator.getInternalBuffer().lastKey() - PoseEstimatorConstants.POSE_BUFFER_SIZE_SECONDS
+				> visionObservation.timestamp();
+		} catch (NoSuchElementException ignored) {
+			return true;
+		}
+	}
+
+	private void addVisionObservation(VisionObservation observation) {
+		Optional<Pose2d> odometryInterpolatedPoseSample = odometryPoseInterpolator.getSample(observation.timestamp());
+		odometryInterpolatedPoseSample.ifPresent(odometryPoseSample -> {
+			Pose2d currentEstimation = PoseEstimationMath
+				.combineVisionToOdometry(observation, odometryPoseSample, estimatedPose, odometryPose, odometryStandardDeviations);
+			estimatedPose = new Pose2d(currentEstimation.getTranslation(), odometryPoseSample.getRotation());
+			estimatedPoseInterpolator.addSample(TimeUtils.getCurrentTimeSeconds(), estimatedPose);
+		});
+	}
+
+	private void addOdometryObservation(OdometryObservation observation) {
+		updateGyroAnglesInLimeLight(observation.gyroAngle());
+		Twist2d twist = kinematics.toTwist2d(latestWheelPositions, observation.wheelsPositions());
+		twist = PoseEstimationMath.addGyroToTwist(twist, observation.gyroAngle(), latestGyroAngle);
+		latestGyroAngle = observation.gyroAngle();
+		latestWheelPositions = observation.wheelsPositions();
+		odometryPose = odometryPose.exp(twist);
+		estimatedPose = estimatedPose.exp(twist);
+		odometryPoseInterpolator.addSample(observation.timestamp(), odometryPose);
+	}
+
+	private void updateGyroAnglesInLimeLight(Rotation2d gyroAngle) {
+		if (gyroAngle != null) {
+			Rotation2d headingWithOffset = gyroAngle.plus(headingOffset);
+			limelightFilterer.updateGyroAngles(new GyroAngleValues(headingWithOffset.getDegrees(), 0, 0, 0, 0, 0));
+		}
+	}
+
+	public void logEstimatedPose() {
+		Logger.recordOutput(super.getLogPath() + "EstimatedPose/", getEstimatedPose());
+	}
+
+	@Override
+	public void subsystemPeriodic() {
+		if (!limelightFilterer.isPoseEstimationCorrect()) {
+			calculateHeadingOffset(latestGyroAngle);
+			resetPoseByLimelight();
+		}
+		updateVision(limelightFilterer.getFilteredVisionObservations());
```

Comment:
1. a sideeffect
2. so why does `updatePoseEstimator` gets limelight filterer outputs... 
(you're run this function twice)

Reply:
what